pkg/controllers/coscheduling/podgroup_controller.go

```diff
@@ -47,16 +66,153 @@ type PodGroupReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.0/pkg/reconcile
 func (r *PodGroupReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	log := log.FromContext(ctx)
+	pg := &schedv1alpha1.PodGroup{}
+	err := r.Get(ctx, req.NamespacedName, pg)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			log.V(5).Info("Pod group has been deleted")
+			return ctrl.Result{}, nil
+		}
+		log.V(3).Error(err, "Unable to retrieve pod group")
+		return ctrl.Result{}, err
+	}
+
+	if pg.Status.Phase == schedv1alpha1.PodGroupFinished ||
+		pg.Status.Phase == schedv1alpha1.PodGroupFailed {
+		return ctrl.Result{}, nil
+	}
+	// If startScheduleTime - createTime > 2days,
+	// do not enqueue again because pod may have been GCed
+	if pg.Status.Scheduled == pg.Spec.MinMember && pg.Status.Running == 0 &&
+		pg.Status.ScheduleStartTime.Sub(pg.CreationTimestamp.Time) > 48*time.Hour {
+		return ctrl.Result{}, nil
+	}
+
+	pgCopy := pg.DeepCopy()
+	podList := &corev1.PodList{}
+	err = r.List(ctx, podList,
+		client.MatchingLabelsSelector{
+			Selector: labels.Set(map[string]string{
+				schedv1alpha1.PodGroupLabel: pgCopy.Name}).AsSelector(),
+		})
+	if err != nil {
+		log.Error(err, "List pods for group failed")
+		return ctrl.Result{}, err
+	}
+	pods := podList.Items
+
+	switch pgCopy.Status.Phase {
+	case "":
+		pgCopy.Status.Phase = schedv1alpha1.PodGroupPending
+	case schedv1alpha1.PodGroupPending:
+		if len(pods) >= int(pg.Spec.MinMember) {
+			pgCopy.Status.Phase = schedv1alpha1.PodGroupPreScheduling
+			fillOccupiedObj(pgCopy, &pods[0])
+		}
+	default:
+		var (
+			running   int32 = 0
+			succeeded int32 = 0
+			failed    int32 = 0
+		)
+		if len(pods) != 0 {
+			for _, pod := range pods {
+				switch pod.Status.Phase {
+				case v1.PodRunning:
+					running++
+				case v1.PodSucceeded:
+					succeeded++
+				case v1.PodFailed:
+					failed++
+				}
+			}
```

Comment:
put this into a function that returns 3 values, something like or similliar:
```go
func getCurrentPodStats(podList v1.PodList) (int32,int32,int32){
}
```

Reply:
make sense, will do it later